### PR TITLE
Admin in db

### PR DESCRIPTION
**Pull request by mcarans**
This has all the changes in the Tables in SQLAlchemy ORM format PR https://github.com/OCHA-DAP/hapi-pipelines/pull/8 as well as code to populate admin tables

This is a very inefficient implementation at present

### CODE REVIEW

File: src/hapi/pipelines/app/pipelines.py [hunk 1]
```diff
@@ -29,6 +30,7 @@ def __init__(
         self.configuration = configuration
         self.session = session
         self.locations = Locations(configuration, session, use_live)
```
Comment (turnerm) [inline src/hapi/pipelines/app/pipelines.py:32]: I know this was done in a different PR, but should we limit locations (and subsequently admin regions) just to HRPs for now?

**Review by turnerm — APPROVED** [review]
Looks good, just some small comments

**Comment by turnerm** [comment]
Actually something I realized: we need to include the empty connector references, i.e. for each admin1, a row in admin2 that has an undefined pcode that references the admin1. And same for admin1 to location.

File: src/hapi/pipelines/utilities/admins.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                InputOptions(encoding="utf-8"),
+            ).cache()
+        except OSError:
+            logger.exception("Download of admin info from dataset failed!")
+            return
+
+        def update_admin_table(desired_adminlevel):
+            for i, row in enumerate(admin_info):
+                adminlevel = row.get("#geo+admin_level")
+                if adminlevel != desired_adminlevel:
```
[the remark this answers is not in the document]
Comment (mcarans) [reply src/hapi/pipelines/utilities/admins.py:35]: Yes good idea

File: src/hapi/pipelines/app/pipelines.py [hunk 3]
```diff
@@ -29,6 +30,7 @@ def __init__(
...
+        self.admins = Admins(session, self.locations)
         self.adminone = AdminLevel(configuration["admin1"], admin_level=1)
```
[the remark this answers is not in the document]
Comment (mcarans) [reply src/hapi/pipelines/app/pipelines.py:34]: The AdminLevel class does the fuzzy matching etc., but I need to have it be set up from the same source as the Admins class (ie. the dataset). This requires changes to HDX Python Country.

File: src/hapi/pipelines/utilities/admins.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                    continue
+                code = row.get("#adm+code")
+                name = row.get("#adm+name")
+                reference_period_start = parse_date(row.get("#date+start"))
+                parent = row.get("#adm+code+parent")
+                if desired_adminlevel == "1":
```
[the remark this answers is not in the document]
Comment (mcarans) [reply src/hapi/pipelines/utilities/admins.py:41]: I am not sure what you have in mind. I'm happy for you to make a PR to structure it better

File: src/hapi/pipelines/utilities/admins.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                else:
+                    admin_ref = self.data[parent]
+                    admin_row = DBAdmin2(
+                        admin1_ref=admin_ref,
+                        code=code,
+                        name=name,
+                        reference_period_start=reference_period_start,
+                    )
+                self.session.add(admin_row)
+                if i % 1000 == 0:
```
[the remark this answers is not in the document]
Comment (mcarans) [reply src/hapi/pipelines/utilities/admins.py:58]: Yes good idea

File: src/hapi/pipelines/utilities/admins.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                    self.session.commit()
+            self.session.commit()
+
+        update_admin_table("1")
+        select(DBAdmin1.id, DBAdmin1.code)
```
[the remark this answers is not in the document]
Comment (mcarans) [reply src/hapi/pipelines/utilities/admins.py:63]: Yes thanks, will remove

File: src/hapi/pipelines/utilities/admins.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        results = self.session.execute(select(DBAdmin1.id, DBAdmin1.code))
+        for result in results:
```
[the remark this answers is not in the document]
Comment (mcarans) [reply src/hapi/pipelines/utilities/admins.py:65]: Good spot.

File: src/hapi/pipelines/utilities/admins.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+            self.data[result[1]] = result[0]
```
[the remark this answers is not in the document]
Comment (mcarans) [reply src/hapi/pipelines/utilities/admins.py:66]: I would think yes, but I defer to @b-j-mills on this

File: src/hapi/pipelines/app/pipelines.py (hunk 1 above)
Comment (mcarans) [reply src/hapi/pipelines/app/pipelines.py:32]: I can make that change in a new PR

Comment (mcarans) [reply src/hapi/pipelines/app/pipelines.py:32]: Actually I'll do it in this one

**Comment by mcarans** [comment]
> Actually something I realized: we need to include the empty connector references, i.e. for each admin1, a row in admin2 that has an undefined pcode that references the admin1. And same for admin1 to location.
> 
> I can take care of it in the population or a separate PR, let me know.

Yes please go ahead. I think we're waiting for @b-j-mills feedback on what to use for XXX in KEN-ABB-XXX.

File: src/hapi/pipelines/app/pipelines.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mcarans) [reply src/hapi/pipelines/app/pipelines.py:34]: I'll look at this in a separate PR

**Pull request merged**